dashboard/app/helpers/aichat_sagemaker_helper.rb

```diff
@@ -1,62 +1,126 @@
 module AichatSagemakerHelper
   ASSISTANT = "assistant"
   USER = "user"
+  SYSTEM = "system"
+  KAREN_PRETEXT = "Edit the following text for spelling and grammar mistakes: "
   INSTRUCTIONS_BEGIN_TOKEN = "[INST]"
   INSTRUCTIONS_END_TOKEN = "[/INST]"
   SENTENCE_BEGIN_TOKEN = "<s>"
   SENTENCE_END_TOKEN = "</s>"
+  CHAT_ML_BEGIN_TOKEN = "<|im_start|>"
+  CHAT_ML_END_TOKEN = "<|im_end|>"
+  NEWLINE = "\n"
   MAX_NEW_TOKENS = 512
   TOP_P = 0.9
+  PIRATE_CUSTOM_STOPPING_STRINGS = ["},"]
+  FINE_TUNED_MODELS = {
+    ARITHMO: "gen-ai-arithmo2-mistral-7b",
+    PIRATE: "gen-ai-mistral-pirate-7b",
+    KAREN: "gen-ai-karen-creative-mistral-7b",
+    BIOMISTRAL: "gen-ai-biomistral-7b"
+  }
 
   def self.create_sagemaker_client
     Aws::SageMakerRuntime::Client.new
   end
 
-  # The instruction-tuned version of Mistral accepts formatted instructions where conversation roles
-  # must start with a user prompt and alternate between user and assistant.
-  # Mistral-7B-Instruction LLM instruction format doc at https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1.
   def self.format_inputs_for_sagemaker_request(aichat_params, stored_messages, new_message)
-    all_messages = [*stored_messages, new_message]
-    inputs = aichat_params[:systemPrompt] + " "
-    inputs += aichat_params[:retrievalContexts].join(" ") if aichat_params[:retrievalContexts]
-    inputs = SENTENCE_BEGIN_TOKEN + wrap_as_instructions(inputs)
-    all_messages.each do |msg|
-      if msg[:role] == USER
-        inputs += wrap_as_instructions(msg[:chatMessageText])
-      elsif msg[:role] == ASSISTANT
-        # Note that each assistant message in the conversation history is followed by
-        # the end-of-sentence token but a begin-of-sentence token is not required.
-        inputs += msg[:chatMessageText] + SENTENCE_END_TOKEN
+    selected_model_id = aichat_params[:selectedModelId]
```

Comment:
Might be time to start writing some tests here, particularly for the base model to ensure we're not regressing anything?

Reply:
Good call! Will do.

Reply:
Hey @bencodeorg - @samantha-code would like to test these fine-tuned models soon so I'll do a quick follow-up to add unit tests. thanks!

Reply:
Follow-up with tests: https://github.com/code-dot-org/code-dot-org/pull/59760